src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -1490,7 +1487,23 @@ private static bool IsTypeSpecForTypicalInstantiation(TypeDesc t)
                     originalMethod = methodOnUnderlyingType;
                 }
 
-                MethodDesc directMethod = constrainedType.TryResolveConstraintMethodApprox(exactType, originalMethod, out forceUseRuntimeLookup);
+                MethodDesc directMethod;
+                if (originalMethod.Signature.IsStatic)
+                {
+                    if (_compilation.AllowCompileTimeStaticVirtualMethodResolution(constrainedType, originalMethod))
+                    {
+                        directMethod = constrainedType.ResolveVariantInterfaceMethodToVirtualMethodOnType(originalMethod);
+                    }
+                    else
+                    {
+                        directMethod = null;
+                    }
+                    forceUseRuntimeLookup = (directMethod == null);
```

Comment:
If we return null because we can't find a result (as there isn't one), and we're not compiling shared generic code, we should produce a CodeGenerationFailedException. I know the current logic in coreclr doesn't fail in the obvious way here, but it does actually reliably fail, and we do have a testcase for it.

Reply:
Hmm, why is it necessary to tear down compilation of the entire method? Wouldn't it be sufficient to just keep the constraint on a particular call and let it be resolved by the runtime?